include/circt/Dialect/HW/HWAttributesNaming.td

```diff
@@ -17,32 +17,71 @@ include "circt/Dialect/HW/HWDialect.td"
 include "mlir/IR/AttrTypeBase.td"
 
 def InnerRefAttr : AttrDef<HWDialect, "InnerRef"> {
-  let summary = "Refer to a name inside a module";
+  let summary = "Refer to a name inside a list of scopes";
   let description = [{
-    This works like a symbol reference, but to a name inside a module.
+    This works like a symbol reference, but to a name inside a list of `InnerSymbolTable`'s.
   }];
   let mnemonic = "innerNameRef";
-  let parameters = (ins "::mlir::FlatSymbolRefAttr":$moduleRef,
-                        "::mlir::StringAttr":$name);
+  let parameters = (ins ArrayRefParameter<"::mlir::StringAttr">:$path);
```

Comment:
Are paths rooted at the (lexical) top?  If so, is the first level required to be a real symbol?  If so, why drop the FlatSymbolRefAttr?
Wouldn't this be something like `optional(?)<FlatSymbolRefAttr>, ArrayRefParameter<FlatInnerSymbolRefAttr>`?

Reply:
Yes, they're always rooted at the top. By "real symbol" do you mean an `mlir::SymbolOp`? If so, then no (as described by point 5). We could still have a `FlatSymbolRefAttr` and interpret that as "just" a name (which may refer to a top-level operation, that may itself implement either an `mlir::SymbolOp` or `hw::InnerSymbolOp`.

Reply:
FlatSymbolRefAttr should always refer to a symbol.  We should coopt an existing typed attribute to refer to a new kind of thing.

Reply:
The root of my original comment is the loss of type information in the path, even if just on the first element.